packages/SwingSet/src/liveslots/liveslots.js

```diff
@@ -827,46 +838,97 @@ function build(
     // create a Promise which callers follow for the result, give it a
     // handler so we can pipeline messages to it, and prepare for the kernel
     // to notify us of its resolution
-    const p = makePipelinablePromise(resultVPID);
+    const pRec = makePipelinablePromise(resultVPID);
+
+    // userspace sees `returnedP` (so that's what we need to register
+    // in slotToVal, and what's what we need to retain with a strong
+    // reference via importedVPIDs), but when dispatch.notify arrives,
+    // we need to fire `pRec.promise` because that's what we've got
+    // the firing controls for
+    importedVPIDs.set(resultVPID, harden({ ...pRec, promise: returnedP }));
+    valToSlot.set(returnedP, resultVPID);
+    slotToVal.set(resultVPID, new WeakRef(returnedP));
 
     // prettier-ignore
     lsdebug(
       `ls.qm send(${JSON.stringify(targetSlot)}, ${legibilizeMethod(prop)}) -> ${resultVPID}`,
     );
     syscall.send(targetSlot, serMethargs, resultVPID);
+
+    // The vpids in the syscall.send might be in A:exportedVPIDs,
+    // B:importedVPIDs, or C:neither. Just after the send(), we are
+    // newly on the hook for following the ones in C:neither. One
+    // option would be to feed all the syscall.send slots to
+    // maybeExportPromise(), which will sort them into A/B/C, then
+    // take everything in C:neither and do a .then on it and add it to
+    // exportedVPIDs. Then we call it a day, and allow all the
+    // resolutions to be delivered in a later turn.
+    //
+    // But instead, we choose the option that says "but many of those
+    // promises might already be resolved", and if there's more than
+    // one, we could amortize some syscall overhead by emitting all
+    // the known resolutions in a 2-or-larger batch, and in this
+    // moment (in this turn) we have a whole list of them that we can
+    // check synchronously.
+    //
+    // To implement this option, the sequence is:
+    // * use W to name the vpids in syscall.send
+    // * feed W into resolutionCollector(), to get 'resolutions'
+    //   * that provides the resolution of any promise in W that is
+    //     known to be resolved, plus any known-resolved promises
+    //     transitively referenced through their resolution data
+    //   * all these resolutions will use the original vpid, which the
+    //     kernel does not currently know about, because the vpid was
+    //     retired earlier, the previous time that promise was
+    //     resolved
+    // * name X the set of vpids resolved in 'resolutions'
+    //   * assert that X vpids are not in exportedVPIDs or importedVPIDs
+    //   * they can only be in X if we remembered the Promise's resolution
+    //   * which means we observed the vpid resolve
+    //   * at that moment of observation, we would have removed it
+    //   * from exportedVPIDs, as we did a syscall.resolve on it
+    // * name Y the set of vpids *referenced* by 'resolutions'
```

Comment:
Use of bullets in this portion of this comment is confusing because some lines are continuations of the lines above them.

Reply:
oops, yeah, my editor wrapped that badly, I'll fix